Updates to Babel 7

ChrisBAshton opened the pull request:
Resolves #1308

**Overall change:** Updates our transpiler version from v6 to v7. Also updates Storybook and Jest (which were previously locked down to older versions as their newer versions required Babel v7).

**Code changes:**

- Updated all updatable Babel dependencies
- Installed [@babel/plugin-transform-modules-commonjs](https://www.npmjs.com/package/@babel/plugin-transform-modules-commonjs) to allow the `module.exports` in `helpers/logger.node.js` and `helpers/logger.web.js`. This was necessary because [Babel 7 dropped use of add-module-exports plugin on Babel packages](https://babeljs.io/docs/en/v7-migration-api#export-changes).
- Updated Storybook dependency to be in sync with Babel v7.
- Updated Jest dependency to be in sync with Babel v7.
- Reverts #1307 (a temporary PR while we were tied to an old - vulnerable - Jest dependency).

NB, `package-lock.json` has big changes, but this is because I've changed a lot of dependencies. It shouldn't be down to an improper setup (I'm using the correct Node and NPM versions):

<img width="695" alt="screen shot 2019-02-18 at 14 44 46" src="https://user-images.githubusercontent.com/5111927/52958341-d34cb080-338b-11e9-822f-20db611b0208.png">

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval
- [x] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

ChrisBAshton committed:
updates babel packages and .babelrc to v7

ChrisBAshton committed:
upgrades storybook to latest version post babel-7 upgrade

ChrisBAshton committed:
upgrades jest to latest version post babel-7 upgrade

ChrisBAshton committed:
reverts #1307 now we've updated Jest dependency

ChrisBAshton commented on .babelrc:
```diff
@@ -21,27 +20,20 @@
-    ],
-    "babel-plugin-syntax-dynamic-import", // allows `await import()`
-    "babel-plugin-dynamic-import-node", // allows `return import()`
-    "transform-object-rest-spread", // allows ...spread notation
```
Replaced by [@babel/plugin-proposal-object-rest-spread](https://www.npmjs.com/package/@babel/plugin-proposal-object-rest-spread).

ChrisBAshton commented on .babelrc:
```diff
@@ -21,27 +20,20 @@
       }
-    ],
-    "babel-plugin-syntax-dynamic-import", // allows `await import()`
-    "babel-plugin-dynamic-import-node", // allows `return import()`
```
Still used, but renamed to `dynamic-import-node`, [according to the docs](https://www.npmjs.com/package/babel-plugin-dynamic-import-node).

ChrisBAshton commented on .babelrc:
```diff
@@ -21,27 +20,20 @@
         "fileName": false // prevent filename forming part of class name (duplication)
       }
-    ],
-    "babel-plugin-syntax-dynamic-import", // allows `await import()`
```
Replaced by [@babel/plugin-syntax-dynamic-import](https://www.npmjs.com/package/@babel/plugin-syntax-dynamic-import).

ChrisBAshton committed:
improves .babelrc annotation

ChrisBAshton commented on .babelrc:
```diff
@@ -21,27 +20,20 @@
-    [
-      // polyfills code without polluting globals
-      // https://stackoverflow.com/a/36821986
-      "transform-runtime",
```
EDIT: have removed without replacement. We have a choice of polyfilling Simorgh or transforming its runtime - we shouldn't do both. Read https://medium.com/@jcse/clearing-up-the-babel-6-ecosystem-c7678a314bf3

~~Replaced by [@babel/plugin-transform-runtime](https://www.npmjs.com/package/@babel/plugin-transform-runtime).~~

ChrisBAshton commented on .babelrc:
```diff
@@ -21,27 +20,20 @@
-      // https://stackoverflow.com/a/36821986
-      "transform-runtime",
-      {
-        "polyfill": false,
```
This is no longer an option in @babel/plugin-transform-runtime (and now defaults to `true` under the hood).

ChrisBAshton commented on package.json:
```diff
@@ -125,7 +125,9 @@
   "devDependencies": {
-    "@storybook/react": "4.0.0-alpha.8",
+    "@babel/plugin-transform-modules-commonjs": "^7.2.0",
+    "@babel/plugin-transform-runtime": "^7.2.0",
```
This is the dev version: https://babeljs.io/docs/en/next/babel-plugin-transform-runtime.html#installation

ChrisBAshton commented on package.json:
```diff
@@ -67,6 +67,13 @@
+    "@babel/polyfill": "^7.2.5",
+    "@babel/preset-env": "^7.3.1",
+    "@babel/preset-react": "^7.0.0",
+    "@babel/runtime": "^7.3.1",
```
This is the prod version: https://babeljs.io/docs/en/next/babel-plugin-transform-runtime.html#installation

bcmn commented:
Looks good after testing behaviours locally.

sareh reviewed (dismissed):
👍 Looks good! Thanks for updating this.

ChrisBAshton committed:
adds storybook:ci, as @storybook/react v4 opens browser by default

ChrisBAshton committed:
merged with latest

ChrisBAshton committed:
adds & annotates 'useBuiltIns' to .babelrc

ChrisBAshton committed:
merge latest

ChrisBAshton committed:
merge SnykLodashMergeBump to fix vuln & allow push

ChrisBAshton committed:
changes useBuiltIns value from 'entry' to 'usage'

Let's discuss the possible values:

* `false` is the default, meaning `import "@babel/polyfill"` imports the entire polyfill module, which is pretty big.
* `"entry"` converts the import into multiple, smaller imports, importing the minimum polyfills required to polyfill the targeted browsers (https://github.com/bbc/simorgh/pull/1312/files#diff-e56633f72ecc521128b3db6586074d2cL7) (e.g. `"last 2 chrome versions"`). See https://babeljs.io/docs/en/babel-preset-env#usebuiltins-entry
* `"usage"` takes it one step further by doing code analysis and only importing polyfills for features that are actually being used. It's classed as "experimental" but errs on the side of "polyfill too much" rather than "too little". In any case, I don't see how it's possible that it would create a bigger bundle than `"entry"` or `false`, so is a good option to choose (and is [the way we're going in Visual Journalism](https://github.com/bbc/news-vj-developer-scaffold/pull/587)). See https://babeljs.io/blog/2018/08/27/7.0.0#automatic-polyfilling-experimental

ChrisBAshton committed:
fixes .babelrc syntax (trailing comma)

ChrisBAshton commented on .babelrc:
```diff
@@ -18,30 +17,26 @@
+        },
+        // converts `import "@babel/polyfill"` into multiple, smaller imports that
+        // polyfills only the features that are used, and only for the targeted browsers
+        "useBuiltIns": "usage"
```
Let's discuss the possible values:

* `false` is the default, meaning `import "@babel/polyfill"` imports the entire polyfill module, which is pretty big.
* `"entry"` converts the import into multiple, smaller imports, importing the minimum polyfills required to polyfill [the targeted browsers](https://github.com/bbc/simorgh/pull/1312/files#diff-e56633f72ecc521128b3db6586074d2cL7) (e.g. `"last 2 chrome versions"`). See https://babeljs.io/docs/en/babel-preset-env#usebuiltins-entry
* `"usage"` takes it one step further by doing code analysis and only importing polyfills for features that are actually being used. It's classed as "experimental" but errs on the side of "polyfill too much" rather than "too little". In any case, I don't see how it's possible that it would create a bigger bundle than `"entry"` or `false`, so is a good option to choose (and is [the way we're going in Visual Journalism](https://github.com/bbc/news-vj-developer-scaffold/pull/587)). See https://babeljs.io/blog/2018/08/27/7.0.0#automatic-polyfilling-experimental

ChrisBAshton committed:
removes '@babel/polyfill' import; not needed with useBuiltIns: usage

ChrisBAshton committed:
debugging Jenkins

ChrisBAshton committed:
removes transform-runtime plugins

We have a choice of polyfilling Simorgh or transforming its runtime - we shouldn't do both. Read https://medium.com/@jcse/clearing-up-the-babel-6-ecosystem-c7678a314bf3

ChrisBAshton committed:
adds babel runtime as prod dependency

Looks as though it's still needed even if not explicitly mentioned in .babelrc

ChrisBAshton committed:
Merge branch 'latest' into babel7

bcmn committed:
Merge branch 'latest' into babel7

bcmn reviewed (approved):
Looks really good -- & testing locally, it looks like there may have already been some benefits to the client bundles from the `useBuiltIns` change when compared to `latest` 👍 

With
![usage](https://user-images.githubusercontent.com/11646957/53637466-212ea780-3c1b-11e9-8f5d-279b3c27e331.png)

Without
![screenshot 2019-03-01 at 12 01 40](https://user-images.githubusercontent.com/11646957/53637485-2ee42d00-3c1b-11e9-9f85-a352ae8a1397.png)

dr3 reviewed (approved):
WOOO :D

sareh committed:
Merge branch 'latest' into babel7

jamesbrumpton commented:
👍
